### PR TITLE
[react] Support anything with `toString()` (such as `TrustedHTML`) as input for `dangerouslySetInnerHTML`.

### PR DESCRIPTION
**Pull request by lgarron**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://developer.mozilla.org/en-US/docs/Web/API/TrustedHTML

This change allows passing `TrustedHTML` to `dangerouslySetInnerHTML` in React. React correctly passes this to `innerHTML` in a way that is compatible with CSP enforcement of the Trusted Types API. We are prototyping a use case for this inside the GitHub codebase.

Since this is a nested property on an interface, I couldn't find a reasonable way to augment the `react` namespace to support this inside a single project. So I thought I'd send this PR to see if we could consider marking the functionality in `@types/react` directly.

### CODE REVIEW

File: types/react/index.d.ts [hunk 1]
```diff
@@ -38,6 +38,7 @@
 import * as CSS from 'csstype';
 import * as PropTypes from 'prop-types';
 import { Interaction as SchedulerInteraction } from 'scheduler/tracing';
+import 'trusted-types';
```
Comment (eps1lon) [inline types/react/index.d.ts:41]: I think we should rather use the same approach we do with the dom related stuff: use an empty interface for `TrustedHTML` and let users take care of the proper type. We shouldn't use the polyfill directly here.

For more guidance check how we test the other DOM interfaces.

**Review by eps1lon — CHANGES_REQUESTED** [review]

**Comment by eps1lon** [comment]
Also: Could you check in which version support for trusted types was added to React?

**Comment by eps1lon** [comment]
I don't think React has actual [support for trusted types](https://github.com/facebook/react/blob/835d9c9f4724b71b429a6b7aaced6da1448e7fb8/packages/shared/ReactFeatureFlags.js#L176) yet.

I think this only works incidentally now because `TrustedHTML` implements `toString()`.

So instead of accepting only `string` we should probably just widen the accepted types to `{ toString(): string }` i.e. there's no need to mention `TrustedHTML` in the typings at all.

File: types/react/index.d.ts (hunk 1 above)
Comment (lgarron) [reply types/react/index.d.ts:41]: Ah, thanks, I thought `trusted-types` were the global types to be used. https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1246 is stalled, so they're not available globally yet. I'll try to take a look.

**Comment by lgarron** [comment]
> I don't think React has actual [support for trusted types](https://github.com/facebook/react/blob/835d9c9f4724b71b429a6b7aaced6da1448e7fb8/packages/shared/ReactFeatureFlags.js#L176) yet.
> 
> I think this only works incidentally now because `TrustedHTML` implements `toString()`.
> 
> So instead of accepting only `string` we should probably just widen the accepted types to `{ toString(): string }` i.e. there's no need to mention `TrustedHTML` in the typings at all.

From what I can tell, React passed the input to `innerHTML` without converting to a string. I think accepting anything with `toString()` makes sense, though, because that string value does get used.

**Comment by eps1lon** [comment]
> From what I can tell, React passed the input to `innerHTML` without converting to a string.

Looks like it: https://codesandbox.io/s/trustedhtml-kfxbhz

> I think accepting anything with `toString()` makes sense, though, because that string value does get used.

Nah, if it really is just assigned we should simply accept the same type as `HTMLElement.prototype.innerHTML`

**Comment by lgarron** [comment]
> Nah, if it really is just assigned we should simply accept the same type as `HTMLElement.prototype.innerHTML`. But then we'd still need to add `TrustedHTML` to https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/bdf9e19b9676d8b1a4ea4fde7407354004e7104a/baselines/dom.generated.d.ts#L8907

That certainly makes sense. If it makes sense to wait on that, would you have any advice for any TypeScript projects that want to pass `TrustedHTML` to `dangerouslySetInnerHTML` today, without suppressing errors?

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react@latest/))
was missing the following properties:
1. unstable_act
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 40151cd5c393ec6547475ef4557ef32950be3bda
</p>

**Comment by eps1lon** [comment]
> would you have any advice for any TypeScript projects that want to pass TrustedHTML to dangerouslySetInnerHTML today, without suppressing errors?

Unfortunately module augmentation doesn't work for existing properties so all you can do is error suppression.

**Comment by lgarron** [comment]
> > would you have any advice for any TypeScript projects that want to pass TrustedHTML to dangerouslySetInnerHTML today, without suppressing errors?
> 
> Unfortunately module augmentation doesn't work for existing properties so all you can do is error suppression.

Do I understand correctly that the best way forward is to advocate for https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1246 to land, and to change `dangerouslySetInnerHTML` to take the type of the `Element.innerHTML` setter from the DOM types?

**Comment by lgarron** [comment]
> change `dangerouslySetInnerHTML` to take the type of the `Element.innerHTML` setter from the DOM types?

Is there a good way to do this? It looks like types from `lib.dom.ts` are not available inside the React definitions right now — would it be worth importing those vs. adding a `toString()` alternative directly to the type?

**Comment by eps1lon** [comment]
> Do I understand correctly that the best way forward is to advocate for https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1246 to land, and to change dangerouslySetInnerHTML to take the type of the Element.innerHTML setter from the DOM types?

Yep

> Is there a good way to do this?

I don't know. One difficulty is how this change will integrate into `react-native`

**Pull request closed without merging**